### ISSUE TITLE
how to nicely handle uncaughtException like the following?

### ISSUE THREAD

**Issue by lyonlai** [Trakkasure/mikronode#52]
uncaughtException { SyntaxError: Expected "!done", "!fatal", "!re", "!trap", ".tag=", [ \t\r\n\x0C\0x00], or [\0x00] but "=" found

I've got this quite often and I'm at 2.0.1. Wondering why is it happen that often. Any idea? routerOS side of thing? or there is issue inside this library.

**Comment by Trakkasure** [comment]
This was due to incomplete parsing code.
Upgrade to 2.3. There have been a lot of bugs resolved since then.
However, there is *one* breaking change. But it it minor.
Let me know if you're still having issues.

**Comment by felipebergamin** [comment]
i'm getting the same error with 2.3.3

**Comment by Trakkasure** [comment]
Can you supply sample code?

**Comment by Trakkasure** [comment]
If you can supply sample code that causes the error, it will help me identify where the error is occurring based on the potential responses from the router.

**Comment by lyonlai** [comment]
my example is just `/ip/dhcp-server/lease/listen`

and when the server is busy it starts causing lots of unhandle exceptions. 

but that's in 2.0.1 though, don't know how much will 2.3.3 help

**Comment by Trakkasure** [comment]
there are a lot of updates in 2.3.3 please test in 2.3.3.
I'll need to simulate a lot of activity, as i do a lease/listen in one of my tests with no errors.

**Comment by felipebergamin** [comment]
hello,

following the examples on README.md i wrote a class to reuse and simplify some code, you can see here: 

https://gist.github.com/felipebergamin/923e9fa86579d01550698d2ed5de8263

in my app.js:

```js
'use strict';

const MkDevice = require('./MkDevice');

const router = new MkDevice('device ip', 'user', 'passwd');

router.getPPPActive()
    .then(console.log);
```


sometimes it works, sometimes returns an error:

```
/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2318
                        throw err;
                        ^
SyntaxError: Expected "!done", "!fatal", "!re", "!trap", ".tag=", [ \t\r\n\x0C\0x00], or end of input but "=" found.
    at peg$buildStructuredError (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:21384:15)
    at Object.peg$parse [as parse] (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:22244:14)
    at MapSubscriber.project (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:474:39)
    at MapSubscriber._next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:13275:36)
    at MapSubscriber.Subscriber.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2213:19)
    at DoSubscriber._next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:11099:31)
    at DoSubscriber.Subscriber.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2213:19)
    at MapSubscriber._next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:13281:27)
    at MapSubscriber.Subscriber.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2213:19)
    at Subject.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:1805:26)
```

using node 8.1.3, routerOS 6.38.35 and 6.27.3, getting error on both versions.

I'm sorry for any English error, I'm brazilian

**Comment by Trakkasure** [comment]
Interesting... so can you turn on debugging? 
Mikronode=require("mikronode");
device=new Mikronode(deviceIp);
device.setDebug(Mikronode.DEBUG);

It should log to console a lot of debugging.
When you get an error, it would be interesting to know the last sentence to have been received.

**Comment by felipebergamin** [comment]
of course!
logging this after a lot of sentences:

```
Channel (chnl)::data flatMap
End of data, nothing left to process
Packet received:  { b: <Buffer >, len: 0, o: [] } <Buffer 12 3d 6c 69 6d 69 74 2d 62 79 74 65 73 2d 6f 75 74 3d 30 0c 3d 72 61 64 69 75 73 3d 74 72 75 65 00 03 21 72 65 0b 2e 74 61 67 3d 63 68 6e 6c 2d 31 0e ... >
Detected end of sentence, posting existing sentence [ '=limit-bytes-out=0', '=radius=true' ]

/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2318
                        throw err;
                        ^
SyntaxError: Expected "!done", "!fatal", "!re", "!trap", ".tag=", [ \t\r\n\x0C\0x00], or [\0x00] but "=" found.
    at peg$buildStructuredError (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:21384:15)
    at Object.peg$parse [as parse] (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:22244:14)
    at MapSubscriber.project (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:474:39)
    at MapSubscriber._next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:13275:36)
    at MapSubscriber.Subscriber.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2213:19)
    at DoSubscriber._next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:11099:31)
    at DoSubscriber.Subscriber.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2213:19)
    at MapSubscriber._next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:13281:27)
    at MapSubscriber.Subscriber.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:2213:19)
    at Subject.next (/home/felipe/Documentos/repos/teste_mikronode/node_modules/mikronode/dist/mikronode.js:1805:26)
```

**Comment by Trakkasure** [comment]
What was before that. A little more back in the history. I'm wondering if
the data is coming in in multiple steams. It looks like the packet ended
before that last piece of data was complete.

**Comment by felipebergamin** [comment]
ok, is this enough?
there's too many ppp clients on router, my terminal is cutting the logs on top... Do you need more?

```
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002A5B',
  '=name=claudia.ferreira@acetech.com.br',
  '=service=pppoe',
  '=caller-id=C4:6E:1F:98:8B:6F',
  '=address=131.0.201.54',
  '=uptime=8h21m43s',
  '=encoding=',
  '=session-id=0x81902A5B',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002A7F',
  '=name=romildo.pereira@acetech.com.br',
  '=service=pppoe',
  '=caller-id=A0:F3:C1:C1:E9:BD',
  '=address=131.0.200.20',
  '=uptime=8h11m35s',
  '=encoding=',
  '=session-id=0x81902A7F',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002AA1',
  '=name=claudio.silva@acetech.com.br',
  '=service=pppoe',
  '=caller-id=30:B5:C2:8E:0B:17',
  '=address=131.0.201.249',
  '=uptime=7h57m22s',
  '=encoding=',
  '=session-id=0x81902AA1',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002D93',
  '=name=lindamira.lara@acetech.com.br',
  '=service=pppoe',
  '=caller-id=18:A6:F7:90:2F:11',
  '=address=131.0.201.22',
  '=uptime=3h22m33s',
  '=encoding=',
  '=session-id=0x81902D93',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002DAC',
  '=name=diogo.siqueira@acetech.com.br',
  '=service=pppoe',
  '=caller-id=30:B5:C2:DF:9A:89',
  '=address=192.168.165.44',
  '=uptime=3h14m6s',
  '=encoding=',
  '=session-id=0x81902DAC',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002DC5',
  '=name=aparecida.silva@acetech.com.br',
  '=service=pppoe',
  '=caller-id=1C:5F:2B:96:95:12',
  '=address=131.0.202.208',
  '=uptime=3h4m24s',
  '=encoding=',
  '=session-id=0x81902DC5',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
Detected end of sentence, posting existing sentence [ '!re',
  '.tag=chnl-1',
  '=.id=*80002DCB',
  '=name=cleide.carvalho@acetech.com.br',
  '=service=pppoe',
  '=caller-id=80:3F:5D:E5:C6:0C',
  '=address=131.0.201.5',
  '=uptime=3h3m17s',
  '=encoding=',
  '=session-id=0x81902DCB',
  '=limit-bytes-in=0',
  '=limit-bytes-out=0',
  '=radius=true' ]
Channel (chnl)::data flatMap
End of data, nothing left to process
Packet received:  { b: <Buffer >, len: 0, o: [] } <Buffer 12 3d 6c 69 6d 69 74 2d 62 79 74 65 73 2d 6f 75 74 3d 30 0c 3d 72 61 64 69 75 73 3d 74 72 75 65 00 03 21 72 65 0b 2e 74 61 67 3d 63 68 6e 6c 2d 31 0e ... >
Detected end of sentence, posting existing sentence [ '=limit-bytes-out=0', '=radius=true' ]
```

**Comment by Trakkasure** [comment]
Cool. That is enough. The open question, not for you to answer but for me, is why there is a leftover sentence at the end where the error occurred.

One more thing... Are there any other channels running?

**Comment by Trakkasure** [comment]
Oh, and what version of routeros is your device running?

**Comment by felipebergamin** [comment]
No, there is only one channel running.
Testing with two CCR 1036 routers, RouterOS 6.38.35 and 6.27.3, this problem occurs on both.

**Comment by Trakkasure** [comment]
Working on this issue. I made a change that allows for catching/resuming on error, but I want to get down to the root cause of why it's receiving an incorrect sentence.

**Comment by felipebergamin** [comment]
Ok, thank you!

If I can help with anything else, please let me know

**Comment by Trakkasure** [comment]
How many items are you getting at one time?
On just a print, I don't get the error.

The pppoe example shouldn't be a big deal.
How big is the PPPoE table? I don't have limits on the stream sizes. I haven't had issues.

Is there any way I can experiment with your system, or a system similar, to figure out what is going on? I cannot seem to reproduce this issue locally.

I am committing the try/catch update code as the next version

**Comment by Trakkasure** [comment]
Ok. I figured it out. I had this issue resolved in the previous version, and I thought it was in this one too.

What is happening, is when the router gets busy, it can post multiple sentences per TCP packet. This packet, when maxing it's max-packet-size, will split between packets. I had accounted for this situation, but right now, it doesn't seem to be working.
Now i know the issue, I can fix it.

**Comment by LBRGeorge** [comment]
@Trakkasure have you fixed? I'm getting this issue as well.

**Comment by Trakkasure** [comment]
I'm working on this at the moment. I will try to get it fixed this weekend.

**Comment by LBRGeorge** [comment]
Nice! Thank you!
If you need help, I'm right here.

**Comment by Trakkasure** [comment]
This issue should now be fixed as of v2.3.4

**Comment by Trakkasure** [comment]
In addition to the update, the parsing was greatly reduces and is now faster. I was able to turn on logging for the firewall on a 100Mb connection for all input and output connections, and the parser kept up.
